// @magenta/music/es5/core/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
null;
exports.DEFAULT_QUARTERS_PER_MINUTE = 120.0;
exports.DEFAULT_STEPS_PER_BAR = 16;
exports.DEFAULT_STEPS_PER_QUARTER = 4;
exports.DEFAULT_STEPS_PER_SECOND = 100;
exports.DEFAULT_VELOCITY = 80;
exports.DEFAULT_PROGRAM = 0;
exports.DEFAULT_TICKS_PER_QUARTER = 220;
exports.DEFAULT_CHANNEL = 0;
exports.DRUM_CHANNEL = 9;
exports.MIN_MIDI_VELOCITY = 0;
exports.MAX_MIDI_VELOCITY = 127;
exports.MIDI_VELOCITIES = exports.MAX_MIDI_VELOCITY - exports.MIN_MIDI_VELOCITY + 1;
exports.NO_CHORD = 'N.C.';
exports.NUM_PITCH_CLASSES = 12;
exports.MIN_MIDI_PITCH = 0;
exports.MAX_MIDI_PITCH = 127;
exports.MIDI_PITCHES = exports.MAX_MIDI_PITCH - exports.MIN_MIDI_PITCH + 1;
exports.MIN_PIANO_PITCH = 21;
exports.MAX_PIANO_PITCH = 108;
exports.MIN_DRUM_PITCH = 35;
exports.MAX_DRUM_PITCH = 81;
exports.MIN_MIDI_PROGRAM = 0;
exports.MAX_MIDI_PROGRAM = 127;
exports.LO_CLICK_PITCH = 89;
exports.HI_CLICK_PITCH = 90;
exports.LO_CLICK_CLASS = 9;
exports.HI_CLICK_CLASS = 10;
