// @magenta/music/es5/core/soundfont.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Tone = require("tone");
var constants = require("./constants");
var Instrument = (function () {
    function Instrument(baseURL) {
        this.FADE_SECONDS = 0.1;
        this.baseURL = baseURL;
        this.buffers = new Tone.Buffers([]);
        this.sourceMap = new Map();
        this.initialized = false;
    }
    Instrument.prototype.initialize = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, fetch(this.baseURL + "/instrument.json")
                            .then(function (response) { return response.json(); })
                            .then(function (spec) {
                            _this.name = spec.name;
                            _this.minPitch = spec.minPitch;
                            _this.maxPitch = spec.maxPitch;
                            _this.durationSeconds = spec.durationSeconds;
                            _this.releaseSeconds = spec.releaseSeconds;
                            _this.percussive = spec.percussive;
                            _this.velocities = spec.velocities;
                            _this.initialized = true;
                        })];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    Instrument.prototype.sampleInfoToName = function (sampleInfo) {
        if (this.velocities) {
            return "p" + sampleInfo.pitch + "_v" + sampleInfo.velocity;
        }
        else {
            return "p" + sampleInfo.pitch;
        }
    };
    Instrument.prototype.sampleNameToURL = function (name) {
        return this.baseURL + "/" + name + ".mp3";
    };
    Instrument.prototype.nearestVelocity = function (velocity) {
        if (!this.velocities) {
            return velocity;
        }
        if (!velocity) {
            velocity = constants.DEFAULT_VELOCITY;
        }
        var bestVelocity = undefined;
        var bestDistance = constants.MIDI_VELOCITIES;
        this.velocities.forEach(function (v) {
            var d = Math.abs(v - velocity);
            if (d < bestDistance) {
                bestVelocity = v;
                bestDistance = d;
            }
        });
        return bestVelocity;
    };
    Instrument.prototype.loadSamples = function (samples) {
        return __awaiter(this, void 0, void 0, function () {
            var nearestSampleNames, uniqueSampleNames, sampleNamesAndURLs;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.initialized) return [3, 2];
                        return [4, this.initialize()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        nearestSampleNames = samples
                            .filter(function (info) {
                            if (info.pitch < _this.minPitch || info.pitch > _this.maxPitch) {
                                console.log("Pitch " + info.pitch + " is outside the valid range for " + _this.name + ", ignoring.");
                                return false;
                            }
                            else {
                                return true;
                            }
                        })
                            .map(function (info) { return _this.sampleInfoToName({
                            pitch: info.pitch,
                            velocity: _this.nearestVelocity(info.velocity)
                        }); });
                        uniqueSampleNames = Array.from(new Set(nearestSampleNames))
                            .filter(function (name) { return !_this.buffers.has(name); });
                        sampleNamesAndURLs = uniqueSampleNames.map(function (name) { return ({ name: name, url: _this.sampleNameToURL(name) }); });
                        if (!(sampleNamesAndURLs.length > 0)) return [3, 4];
                        sampleNamesAndURLs.forEach(function (nameAndURL) { return _this.buffers.add(nameAndURL.name, nameAndURL.url); });
                        return [4, new Promise(function (resolve) { return Tone.Buffer.on('load', resolve); })];
                    case 3:
                        _a.sent();
                        console.log("Loaded samples for " + this.name + ".");
                        _a.label = 4;
                    case 4: return [2];
                }
            });
        });
    };
    Instrument.prototype.playNote = function (pitch, velocity, startTime, duration, output) {
        var buffer = this.getBuffer(pitch, velocity);
        if (duration > this.durationSeconds) {
            console.log("Requested note duration longer than sample duration: " + duration + " > " + this.durationSeconds);
        }
        var source = new Tone.BufferSource(buffer).connect(output);
        source.start(startTime, 0, undefined, 1, 0);
        if (!this.percussive && duration < this.durationSeconds) {
            var releaseSource = new Tone.BufferSource(buffer).connect(output);
            source.stop(startTime + duration + this.FADE_SECONDS, this.FADE_SECONDS);
            releaseSource.start(startTime + duration, this.durationSeconds, undefined, 1, this.FADE_SECONDS);
        }
    };
    Instrument.prototype.playNoteDown = function (pitch, velocity, output) {
        var buffer = this.getBuffer(pitch, velocity);
        var source = new Tone.BufferSource(buffer).connect(output);
        source.start(0, 0, undefined, 1, 0);
        if (this.sourceMap.has(pitch)) {
            this.sourceMap.get(pitch).stop(Tone.now() + this.FADE_SECONDS, this.FADE_SECONDS);
        }
        this.sourceMap.set(pitch, source);
    };
    Instrument.prototype.playNoteUp = function (pitch, velocity, output) {
        if (!this.sourceMap.has(pitch)) {
            return;
        }
        var buffer = this.getBuffer(pitch, velocity);
        var releaseSource = new Tone.BufferSource(buffer).connect(output);
        releaseSource.start(0, this.durationSeconds, undefined, 1, this.FADE_SECONDS);
        this.sourceMap.get(pitch).stop(Tone.now() + this.FADE_SECONDS, this.FADE_SECONDS);
        this.sourceMap.delete(pitch);
    };
    Instrument.prototype.getBuffer = function (pitch, velocity) {
        if (!this.initialized) {
            throw new Error('Instrument is not initialized.');
        }
        if (pitch < this.minPitch || pitch > this.maxPitch) {
            console.log("Pitch " + pitch + " is outside the valid range for " + this.name + " (" + this.minPitch + "-" + this.maxPitch + ")");
            return;
        }
        var name = this.sampleInfoToName({ pitch: pitch, velocity: this.nearestVelocity(velocity) });
        if (!this.buffers.has(name)) {
            throw new Error("Buffer not found for " + this.name + ": " + name);
        }
        var buffer = this.buffers.get(name);
        if (!buffer.loaded) {
            throw new Error("Buffer not loaded for " + this.name + ": " + name);
        }
        return buffer;
    };
    return Instrument;
}());
exports.Instrument = Instrument;
var SoundFont = (function () {
    function SoundFont(baseURL) {
        this.baseURL = baseURL;
        this.instruments = new Map();
        this.initialized = false;
    }
    SoundFont.prototype.initialize = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, fetch(this.baseURL + "/soundfont.json")
                            .then(function (response) { return response.json(); })
                            .then(function (spec) {
                            _this.name = spec.name;
                            for (var program in spec.instruments) {
                                var url = _this.baseURL + "/" + spec.instruments[program];
                                _this.instruments.set(program === 'drums' ? 'drums' : +program, new Instrument(url));
                            }
                            _this.initialized = true;
                        })];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    SoundFont.prototype.loadSamples = function (samples) {
        return __awaiter(this, void 0, void 0, function () {
            var instrumentSamples;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.initialized) return [3, 2];
                        return [4, this.initialize()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        instrumentSamples = new Map();
                        samples.forEach(function (info) {
                            info.isDrum = info.isDrum || false;
                            info.program = info.program || 0;
                            var instrument = info.isDrum ? 'drums' : info.program;
                            var sampleInfo = { pitch: info.pitch, velocity: info.velocity };
                            if (!instrumentSamples.has(instrument)) {
                                if (!_this.instruments.has(instrument)) {
                                    console.log("No instrument in " + _this.name + " for: program=" + info.program + ", isDrum=" + info.isDrum);
                                }
                                else {
                                    instrumentSamples.set(instrument, [sampleInfo]);
                                }
                            }
                            else {
                                instrumentSamples.get(instrument).push(sampleInfo);
                            }
                        });
                        return [4, Promise.all(Array.from(instrumentSamples.keys())
                                .map(function (info) { return _this.instruments.get(info).loadSamples(instrumentSamples.get(info)); }))];
                    case 3:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    SoundFont.prototype.playNote = function (pitch, velocity, startTime, duration, program, isDrum, output) {
        if (program === void 0) { program = 0; }
        if (isDrum === void 0) { isDrum = false; }
        var instrument = isDrum ? 'drums' : program;
        if (!this.initialized) {
            throw new Error('SoundFont is not initialized.');
        }
        if (!this.instruments.has(instrument)) {
            console.log("No instrument in " + this.name + " for: program=" + program + ", isDrum=" + isDrum);
            return;
        }
        this.instruments.get(instrument)
            .playNote(pitch, velocity, startTime, duration, output);
    };
    SoundFont.prototype.playNoteDown = function (pitch, velocity, program, isDrum, output) {
        if (program === void 0) { program = 0; }
        if (isDrum === void 0) { isDrum = false; }
        var instrument = isDrum ? 'drums' : program;
        if (!this.initialized) {
            throw new Error('SoundFont is not initialized.');
        }
        if (!this.instruments.has(instrument)) {
            console.log("No instrument in " + this.name + " for: program=" + program + ", isDrum=" + isDrum);
            return;
        }
        this.instruments.get(instrument).playNoteDown(pitch, velocity, output);
    };
    SoundFont.prototype.playNoteUp = function (pitch, velocity, program, isDrum, output) {
        if (program === void 0) { program = 0; }
        if (isDrum === void 0) { isDrum = false; }
        var instrument = isDrum ? 'drums' : program;
        if (!this.initialized) {
            throw new Error('SoundFont is not initialized.');
        }
        if (!this.instruments.has(instrument)) {
            console.log("No instrument in " + this.name + " for: program=" + program + ", isDrum=" + isDrum);
            return;
        }
        this.instruments.get(instrument).playNoteUp(pitch, velocity, output);
    };
    return SoundFont;
}());
exports.SoundFont = SoundFont;
